field/src/coset.rs

```diff
@@ -0,0 +1,252 @@
+        // which would come at an unnecessary performance cost in the vast
+        // majority of use cases.
+        assert!(
+            index < 1 << self.log_size,
+            "index must be less than the size of the coset. \
+            Consider passing the equivalent index % (1 << log_size) instead"
+        );
+        self.shift * self.generator_exp(index)
+    }
+
+    // Internal function which computes `generator^exp`. It uses the
+    // square-and-multiply algorithm with the caveat that squares of the
+    // generator are queried from the field (which typically should have them
+    // stored), i. e. rather "fetch-and-multiply"
+    fn generator_exp(&mut self, exp: usize) -> F {
+        let mut gen_power = F::ONE;
+        let mut exp = exp;
+        let mut i = self.log_size();
+
+        while exp > 0 {
+            if exp & 1 != 0 {
+                gen_power *= F::two_adic_generator(i);
+            }
+            exp >>= 1;
+
+            // This cannot be added to the loop condition cleanly because
+            // casting it to an isize would give a different range than usize
+            if i == 0 {
+                break;
+            }
+
+            i -= 1;
+        }
+
+        gen_power
+    }
+
+    /// Returns an iterator over the elements of the coset in the canonical order
+    /// "shift * generator^0, shift * generator^1, ...,
+    /// shift * generator^(2^log_size - 1)`.
+    pub fn iter(&self) -> Take<Powers<F>> {
+        self.generator()
+            .shifted_powers(self.shift)
+            .take(1 << self.log_size)
+    }
+}
+
+/// This tests the equality of the two cosets as mathematical objects, that is:
+/// whether they contain the exact same elements (not necessarily in the same
+/// order) or not.
+impl<F: TwoAdicField> PartialEq for TwoAdicMultiplicativeCoset<F> {
+    fn eq(&self, other: &Self) -> bool {
+        if self.generator() == other.generator() && self.shift == other.shift {
+            return true;
+        }
+
+        // The first equality assumes generators are chosen canonically (as
+        // ensured by the TwoAdicField interface). If not, one could simply
+        // assert self.generator has the same order as other.generator (this is
+        // enough by the cyclicity of the group of units)
+        self.generator() == other.generator() && other.contains(self.shift)
+    }
+}
+
+/// This tests the equality of the two cosets as mathematical objects, that is:
+/// whether they contain the exact same elements (not necessarily in the same
+/// order) or not.
+///
+// This falls back to PartialEq
+impl<F: TwoAdicField> Eq for TwoAdicMultiplicativeCoset<F> {}
+
+impl<F: TwoAdicField> IntoIterator for TwoAdicMultiplicativeCoset<F> {
+    type Item = F;
+    type IntoIter = Take<Powers<F>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.generator()
+            .shifted_powers(self.shift)
+            .take(1 << self.log_size)
+    }
+}
```

Comment:
Is this just a convenience method so you don't need to call `.iter()`? I'm not sure if we need this either but if we do want it we can simplify the code slightly:
```suggestion
impl<F: TwoAdicField> IntoIterator for TwoAdicMultiplicativeCoset<F> {
    type Item = F;
    type IntoIter = Take<Powers<F>>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}
```

Reply:
Indeed, it was just a convenience method to be able to write things like `for element in coset {...}`. Thanks for your clean suggestion, we have incorporated it.